pinot-tools/src/main/java/com/linkedin/pinot/tools/perf/QueryRunner.java

```diff
@@ -0,0 +1,586 @@
+      LOGGER.info("Time Passed: {}ms, Average QPS: {}, Average Broker Time: {}ms, Average Client Time: {}ms.",
+          timePassed, numQueriesExecuted.get() / ((double) timePassed / MILLIS_PER_SECOND),
+          totalBrokerTime.get() / (double) numQueriesExecuted.get(),
+          totalClientTime.get() / (double) numQueriesExecuted.get());
+      numReportIntervals++;
+      if (numReportIntervals == numIntervalsToReportClientTimeStatistics) {
+        reportClientTimeStatistics(clientTimeStatistics);
+        numReportIntervals = 0;
+      }
+    }
+  }
+
+  /**
+   * Use multiple threads to run query at a target QPS.
+   * <p>Use a concurrent linked queue to buffer the queries to be sent. Use the main thread to insert queries into the
+   * queue at the target QPS, and start <code>numThreads</code> worker threads to fetch queries from the queue and send
+   * them.
+   * <p>The main thread is responsible for collecting and logging the statistic information periodically.
+   * <p>Queries are picked randomly from the query file.
+   * <p>Query runner will run forever or until too many queries blocked in the queue.
+   *
+   * @param conf perf benchmark driver config.
+   * @param queryFile query file.
+   * @param numThreads number of threads sending queries.
+   * @param startQPS start QPS (target QPS).
+   * @param reportIntervalMs report interval in milliseconds.
+   * @param numIntervalsToReportClientTimeStatistics number of intervals to report detailed client time statistics.
+   * @throws Exception
+   */
+  @SuppressWarnings("InfiniteLoopStatement")
+  public static void targetQPSQueryRunner(PerfBenchmarkDriverConf conf, String queryFile, int numThreads,
+      double startQPS, int reportIntervalMs, int numIntervalsToReportClientTimeStatistics)
+      throws Exception {
+    Random random = new Random();
+
+    List<String> queries;
+    try (FileInputStream input = new FileInputStream(new File(queryFile))) {
+      queries = IOUtils.readLines(input);
+    }
+    int numQueries = queries.size();
+
+    PerfBenchmarkDriver driver = new PerfBenchmarkDriver(conf);
+    AtomicInteger numQueriesExecuted = new AtomicInteger(0);
+    AtomicLong totalBrokerTime = new AtomicLong(0L);
+    AtomicLong totalClientTime = new AtomicLong(0L);
+    DescriptiveStatistics clientTimeStatistics = new DescriptiveStatistics();
+    ConcurrentLinkedQueue<String> queryQueue = new ConcurrentLinkedQueue<>();
+
+    ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+    for (int i = 0; i < numThreads; i++) {
+      executorService.submit(
+          new Worker(driver, numQueriesExecuted, totalBrokerTime, totalClientTime, clientTimeStatistics, queryQueue));
+    }
+    executorService.shutdown();
+
+    int intervalMs = (int) (MILLIS_PER_SECOND / startQPS);
+    int queueLengthThreshold = Math.max(20, (int) startQPS);
+    long startTime = System.currentTimeMillis();
+    int numReportIntervals = 0;
+    while (true) {
```

Comment:
Pls add one, maybe numQueries param. Don't want to kill the tool. 

I'm running this tool many many times with different setup each time to evaluate scale limits. Killing  is not a good option.